{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "17842f9e-2daf-4520-a42d-67aa0289231b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import cerberus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5d9fa064-c2e7-43f2-9525-566ff4549bf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6a3894e-72ca-4cd9-8e16-9e86e6399033",
   "metadata": {},
   "outputs": [],
   "source": [
    "iq_struct = 'IsoID_TranscriptID_TSS_Intron_PolyA'\n",
    "ofile = 'scisoseq.gtf'\n",
    "gene_meta = '/Users/fairliereese/mortazavi_lab/data/mousewg/refs/gencode_vM21_gene_metadata.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4e0309e4-209d-40d3-9cd5-0df13f77b229",
   "metadata": {},
   "outputs": [],
   "source": [
    "isoquant_struct_to_gtf(iq_struct, \n",
    "                       gene_meta,\n",
    "                       ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf5a430f-6f4c-4f10-a2b9-d685737fb8d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('IsoID_TranscriptID_TSS_Intron_PolyA', sep='\\t', header=None,\n",
    "                names=['transcript_id', 'original_transcript_id', 'tss', 'ic', 'tes'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "92e8f29c-c0b8-43c4-aad9-266664a39bc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>ic</th>\n",
       "      <th>tes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001.4_1</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108107365_108107436_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000001.4_2</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108108143_108108180_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000001.4_3</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108107622_108107622_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000001.4_4</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108108709_108108784_-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id original_transcript_id                         tss  \\\n",
       "0  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "1  ENSMUSG00000000001.4_1   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "2  ENSMUSG00000000001.4_2   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "3  ENSMUSG00000000001.4_3   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "4  ENSMUSG00000000001.4_4   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "\n",
       "                                                  ic  \\\n",
       "0  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "1  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "2  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "3  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "4  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "\n",
       "                          tes  \n",
       "0  chr3_108107270_108107306_-  \n",
       "1  chr3_108107365_108107436_-  \n",
       "2  chr3_108108143_108108180_-  \n",
       "3  chr3_108107622_108107622_-  \n",
       "4  chr3_108108709_108108784_-  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d9605e2f-832b-42ba-9747-fe5ad082b236",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull chromosome and strand out from tss\n",
    "df[['Chromosome', 'Strand']] = df.tss.str.split('_', expand=True)[[0,3]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "253971c4-603e-46c5-9383-6823a4e73c86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>ic</th>\n",
       "      <th>tes</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001.4_1</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108107365_108107436_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000001.4_2</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108108143_108108180_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000001.4_3</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108107622_108107622_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000001.4_4</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>;%;chr3_108109317_108109402_-;%;chr3_108109613...</td>\n",
       "      <td>chr3_108108709_108108784_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id original_transcript_id                         tss  \\\n",
       "0  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "1  ENSMUSG00000000001.4_1   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "2  ENSMUSG00000000001.4_2   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "3  ENSMUSG00000000001.4_3   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "4  ENSMUSG00000000001.4_4   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "\n",
       "                                                  ic  \\\n",
       "0  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "1  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "2  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "3  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "4  ;%;chr3_108109317_108109402_-;%;chr3_108109613...   \n",
       "\n",
       "                          tes Chromosome Strand  \n",
       "0  chr3_108107270_108107306_-       chr3      -  \n",
       "1  chr3_108107365_108107436_-       chr3      -  \n",
       "2  chr3_108108143_108108180_-       chr3      -  \n",
       "3  chr3_108107622_108107622_-       chr3      -  \n",
       "4  chr3_108108709_108108784_-       chr3      -  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Strand == '-'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "84108ecb-e34b-48f2-b4e8-5616900ec4f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['coords'] = df.ic.str.split(';%;')\n",
    "df.drop('ic', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "df4a5d74-1ed9-4f5a-bbdc-6654fb82fba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.explode('coords', ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ef7f5418-1df9-4f37-94c5-70ffc0a2523c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove weird blank entries \n",
    "df = df.loc[df.coords != '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "06f16075-f609-4fce-bca3-729b3d0f87b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get start and stop of each intron\n",
    "df[['Start', 'End']] = df.coords.str.split('_', expand=True)[[1,2]]\n",
    "df[['Start', 'End']] = df[['Start', 'End']].astype(int)\n",
    "df.drop('coords', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2b8edfee-1391-4018-a15c-4addd2840521",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121237352</td>\n",
       "      <td>121238418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121238518</td>\n",
       "      <td>121242500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121242645</td>\n",
       "      <td>121244593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121244727</td>\n",
       "      <td>121244935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121245071</td>\n",
       "      <td>121245337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121245457</td>\n",
       "      <td>121246925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121247056</td>\n",
       "      <td>121249128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121249193</td>\n",
       "      <td>121250292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121250431</td>\n",
       "      <td>121250759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121250918</td>\n",
       "      <td>121252680</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id original_transcript_id  \\\n",
       "183  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "184  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "185  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "186  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "187  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "188  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "189  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "190  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "191  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "192  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "\n",
       "                             tss                          tes Chromosome  \\\n",
       "183  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "184  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "185  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "186  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "187  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "188  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "189  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "190  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "191  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "192  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "\n",
       "    Strand      Start        End  \n",
       "183      +  121237352  121238418  \n",
       "184      +  121238518  121242500  \n",
       "185      +  121242645  121244593  \n",
       "186      +  121244727  121244935  \n",
       "187      +  121245071  121245337  \n",
       "188      +  121245457  121246925  \n",
       "189      +  121247056  121249128  \n",
       "190      +  121249193  121250292  \n",
       "191      +  121250431  121250759  \n",
       "192      +  121250918  121252680  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(df.loc[df.transcript_id=='ENSMUSG00000000056.7_0'].index))\n",
    "df.loc[df.transcript_id=='ENSMUSG00000000056.7_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e9271b9f-369f-47e1-8f06-191ff234810a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# increment / decrement coords to turn to exons\n",
    "df.Start = df.Start-1\n",
    "df.End = df.End+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "64c5ce8e-986a-48f5-ad22-1bff6ddc0d07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109316</td>\n",
       "      <td>108109403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109612</td>\n",
       "      <td>108111935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112088</td>\n",
       "      <td>108112473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112602</td>\n",
       "      <td>108115763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108115891</td>\n",
       "      <td>108118301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108118458</td>\n",
       "      <td>108123542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123683</td>\n",
       "      <td>108123795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123837</td>\n",
       "      <td>108145888</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id original_transcript_id                         tss  \\\n",
       "1  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "2  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "3  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "4  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "5  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "6  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "7  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "8  ENSMUSG00000000001.4_0   ENSMUST00000000001.4  chr3_108146081_108146150_-   \n",
       "\n",
       "                          tes Chromosome Strand      Start        End  \n",
       "1  chr3_108107270_108107306_-       chr3      -  108109316  108109403  \n",
       "2  chr3_108107270_108107306_-       chr3      -  108109612  108111935  \n",
       "3  chr3_108107270_108107306_-       chr3      -  108112088  108112473  \n",
       "4  chr3_108107270_108107306_-       chr3      -  108112602  108115763  \n",
       "5  chr3_108107270_108107306_-       chr3      -  108115891  108118301  \n",
       "6  chr3_108107270_108107306_-       chr3      -  108118458  108123542  \n",
       "7  chr3_108107270_108107306_-       chr3      -  108123683  108123795  \n",
       "8  chr3_108107270_108107306_-       chr3      -  108123837  108145888  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id=='ENSMUSG00000000056.7_0']\n",
    "df.loc[df.transcript_id=='ENSMUSG00000000001.4_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6abfcbc-1171-40cb-be90-12dace364f70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# melt to get coords individually\n",
    "df = df.melt(id_vars=['transcript_id', 'original_transcript_id',\n",
    "                 'tss', 'tes', 'Chromosome', 'Strand'],\n",
    "                 value_vars=['Start', 'End'],\n",
    "                 value_name='Coord')\n",
    "df.drop('variable', axis=1, inplace=True)\n",
    "df.Coord = df.Coord.astype(int)\n",
    "\n",
    "fwd, rev = cerberus.get_stranded_gtf_dfs(df)\n",
    "fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, True], inplace=True)\n",
    "rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, False], inplace=True)\n",
    "\n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "57c816bd-49d4-459f-9c36-bb631772c9e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121237351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636555</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121238419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121238517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636556</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121242501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121242644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636557</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121244594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121244726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636558</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121244936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121245070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636559</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121245338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121245456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636560</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121246926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121247055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636561</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121249129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121249192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636562</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121250293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121250430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636563</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121250760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121250917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636564</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11_121237217_121237289_+</td>\n",
       "      <td>chr11_121254530_121254583_+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>121252681</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript_id original_transcript_id  \\\n",
       "159     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636555  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "160     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636556  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "161     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636557  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "162     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636558  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "163     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636559  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "164     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636560  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "165     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636561  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "166     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636562  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "167     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636563  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "168     ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "636564  ENSMUSG00000000056.7_0   ENSMUST00000103015.3   \n",
       "\n",
       "                                tss                          tes Chromosome  \\\n",
       "159     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636555  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "160     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636556  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "161     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636557  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "162     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636558  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "163     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636559  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "164     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636560  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "165     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636561  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "166     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636562  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "167     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636563  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "168     chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "636564  chr11_121237217_121237289_+  chr11_121254530_121254583_+      chr11   \n",
       "\n",
       "       Strand      Coord  \n",
       "159         +  121237351  \n",
       "636555      +  121238419  \n",
       "160         +  121238517  \n",
       "636556      +  121242501  \n",
       "161         +  121242644  \n",
       "636557      +  121244594  \n",
       "162         +  121244726  \n",
       "636558      +  121244936  \n",
       "163         +  121245070  \n",
       "636559      +  121245338  \n",
       "164         +  121245456  \n",
       "636560      +  121246926  \n",
       "165         +  121247055  \n",
       "636561      +  121249129  \n",
       "166         +  121249192  \n",
       "636562      +  121250293  \n",
       "167         +  121250430  \n",
       "636563      +  121250760  \n",
       "168         +  121250917  \n",
       "636564      +  121252681  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fwd.loc[fwd.transcript_id=='ENSMUSG00000000056.7_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "11eea56e-e064-440b-bc87-9758a75df7c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get start / end coords of tss / tes regions\n",
    "df[['tss_Start', 'tss_End']] = df.tss.str.split('_', expand=True)[[1,2]].astype(int)\n",
    "df[['tes_Start', 'tes_End']] = df.tes.str.split('_', expand=True)[[1,2]].astype(int)\n",
    "\n",
    "fwd, rev = cerberus.get_stranded_gtf_dfs(df)\n",
    "\n",
    "# loop through strands and ends \n",
    "ends = pd.DataFrame()\n",
    "for mode in ['tss', 'tes']:\n",
    "    for strand, temp in zip(['+', '-'], [fwd, rev]):\n",
    "        end_cols = ['{}_Start'.format(mode),\n",
    "                    '{}_End'.format(mode)]\n",
    "        keep_cols = ['transcript_id', 'original_transcript_id',\n",
    "                     'Chromosome', 'Strand']\n",
    "        keep_cols += end_cols\n",
    "        temp = temp[keep_cols]\n",
    "        old_end, new_end, func = cerberus.get_update_ends_settings(strand, mode)\n",
    "        temp['Coord'] = temp[end_cols].apply(func, axis=1)\n",
    "        # print(len(temp.loc[temp['Coord'].isnull()].index))\n",
    "        temp.drop(end_cols, axis=1, inplace=True)\n",
    "        temp.drop_duplicates(inplace=True)\n",
    "        ends = pd.concat([ends, temp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "02a1e752-f90c-442e-8de2-da04d0623b08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108146150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108107270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id original_transcript_id Chromosome Strand      Coord\n",
       "0  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -  108146150\n",
       "0  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -  108107270"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ends.loc[ends.transcript_id=='ENSMUSG00000000056.7_0']\n",
    "ends.loc[ends.transcript_id=='ENSMUSG00000000001.4_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3db6a3fb-75fb-4c71-aa74-3223d5b786b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coord</th>\n",
       "      <th>tss_Start</th>\n",
       "      <th>tss_End</th>\n",
       "      <th>tes_Start</th>\n",
       "      <th>tes_End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109316</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109612</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112088</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112602</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108115891</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108118458</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123683</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123837</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636396</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109403</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636397</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108111935</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636398</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112473</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636399</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108115763</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636400</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108118301</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636401</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123542</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636402</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123795</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636403</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3_108146081_108146150_-</td>\n",
       "      <td>chr3_108107270_108107306_-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108145888</td>\n",
       "      <td>108146081</td>\n",
       "      <td>108146150</td>\n",
       "      <td>108107270</td>\n",
       "      <td>108107306</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript_id original_transcript_id  \\\n",
       "0       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "1       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "2       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "3       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "4       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "5       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "6       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "7       ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636396  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636397  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636398  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636399  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636400  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636401  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636402  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "636403  ENSMUSG00000000001.4_0   ENSMUST00000000001.4   \n",
       "\n",
       "                               tss                         tes Chromosome  \\\n",
       "0       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "1       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "2       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "3       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "4       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "5       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "6       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "7       chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636396  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636397  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636398  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636399  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636400  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636401  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636402  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "636403  chr3_108146081_108146150_-  chr3_108107270_108107306_-       chr3   \n",
       "\n",
       "       Strand      Coord  tss_Start    tss_End  tes_Start    tes_End  \n",
       "0           -  108109316  108146081  108146150  108107270  108107306  \n",
       "1           -  108109612  108146081  108146150  108107270  108107306  \n",
       "2           -  108112088  108146081  108146150  108107270  108107306  \n",
       "3           -  108112602  108146081  108146150  108107270  108107306  \n",
       "4           -  108115891  108146081  108146150  108107270  108107306  \n",
       "5           -  108118458  108146081  108146150  108107270  108107306  \n",
       "6           -  108123683  108146081  108146150  108107270  108107306  \n",
       "7           -  108123837  108146081  108146150  108107270  108107306  \n",
       "636396      -  108109403  108146081  108146150  108107270  108107306  \n",
       "636397      -  108111935  108146081  108146150  108107270  108107306  \n",
       "636398      -  108112473  108146081  108146150  108107270  108107306  \n",
       "636399      -  108115763  108146081  108146150  108107270  108107306  \n",
       "636400      -  108118301  108146081  108146150  108107270  108107306  \n",
       "636401      -  108123542  108146081  108146150  108107270  108107306  \n",
       "636402      -  108123795  108146081  108146150  108107270  108107306  \n",
       "636403      -  108145888  108146081  108146150  108107270  108107306  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df.loc[df.transcript_id=='ENSMUSG00000000056.7_0']\n",
    "df.loc[df.transcript_id=='ENSMUSG00000000001.4_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a3c07e75-4827-48bd-80a0-1a2fbfdcbd19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "121237217 < 121237289"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d4f9862c-e853-4449-add2-cbba7e6be1c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "108146150.0>108107270"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "faadb497-9ca1-4446-b263-a1a46e2a4252",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "108107270.0<108107306"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ae7164de-843e-413b-9274-4509dd561bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnecessary columns\n",
    "df.drop(['tss', 'tes', 'tss_Start', 'tes_Start', 'tss_End', 'tes_End'], \n",
    "        axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b122900a-131e-44ca-b4bd-2ae25161d7aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate and then sort based on strand\n",
    "df = pd.concat([df, ends])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d2d2e911-4173-4f7b-a2de-edb0ab63cdc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "fwd, rev = cerberus.get_stranded_gtf_dfs(df)\n",
    "fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, True], inplace=True)\n",
    "rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, False], inplace=True)\n",
    "df = pd.concat([fwd, rev])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "85aa70c3-e2b2-4c4a-af0c-4a56bb06b397",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108146150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636403</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108145888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636402</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636401</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108123542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108118458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636400</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108118301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108115891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636399</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108115763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636398</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108112088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636397</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108111935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636396</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108109316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4_0</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>chr3</td>\n",
       "      <td>-</td>\n",
       "      <td>108107270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript_id original_transcript_id Chromosome Strand  \\\n",
       "0       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636403  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "7       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636402  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "6       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636401  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "5       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636400  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "4       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636399  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "3       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636398  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "2       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636397  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "1       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "636396  ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "0       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "0       ENSMUSG00000000001.4_0   ENSMUST00000000001.4       chr3      -   \n",
       "\n",
       "            Coord  \n",
       "0       108146150  \n",
       "636403  108145888  \n",
       "7       108123837  \n",
       "636402  108123795  \n",
       "6       108123683  \n",
       "636401  108123542  \n",
       "5       108118458  \n",
       "636400  108118301  \n",
       "4       108115891  \n",
       "636399  108115763  \n",
       "3       108112602  \n",
       "636398  108112473  \n",
       "2       108112088  \n",
       "636397  108111935  \n",
       "1       108109612  \n",
       "636396  108109403  \n",
       "0       108109316  \n",
       "0       108107270  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id=='ENSMUSG00000000056.7_0']\n",
    "df.loc[df.transcript_id=='ENSMUSG00000000001.4_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b2fdefa0-2da6-4ffe-a822-ce43cbaa2958",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene id\n",
    "df['gene_id'] = df.transcript_id.str.split('_', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "65ba8990-78ee-413b-8704-c9136d8fd249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coord</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26947</th>\n",
       "      <td>ENSMUSG00000003135.15_0</td>\n",
       "      <td>ENSMUST00000003219.13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>39535748</td>\n",
       "      <td>ENSMUSG00000003135.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26947</th>\n",
       "      <td>ENSMUSG00000003135.15_0</td>\n",
       "      <td>ENSMUST00000003219.13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>39536337</td>\n",
       "      <td>ENSMUSG00000003135.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663343</th>\n",
       "      <td>ENSMUSG00000003135.15_0</td>\n",
       "      <td>ENSMUST00000003219.13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>39537416</td>\n",
       "      <td>ENSMUSG00000003135.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26948</th>\n",
       "      <td>ENSMUSG00000003135.15_0</td>\n",
       "      <td>ENSMUST00000003219.13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>39537580</td>\n",
       "      <td>ENSMUSG00000003135.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663344</th>\n",
       "      <td>ENSMUSG00000003135.15_0</td>\n",
       "      <td>ENSMUST00000003219.13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>39538271</td>\n",
       "      <td>ENSMUSG00000003135.15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  transcript_id original_transcript_id Chromosome Strand  \\\n",
       "26947   ENSMUSG00000003135.15_0  ENSMUST00000003219.13       chr1      +   \n",
       "26947   ENSMUSG00000003135.15_0  ENSMUST00000003219.13       chr1      +   \n",
       "663343  ENSMUSG00000003135.15_0  ENSMUST00000003219.13       chr1      +   \n",
       "26948   ENSMUSG00000003135.15_0  ENSMUST00000003219.13       chr1      +   \n",
       "663344  ENSMUSG00000003135.15_0  ENSMUST00000003219.13       chr1      +   \n",
       "\n",
       "           Coord                gene_id  \n",
       "26947   39535748  ENSMUSG00000003135.15  \n",
       "26947   39536337  ENSMUSG00000003135.15  \n",
       "663343  39537416  ENSMUSG00000003135.15  \n",
       "26948   39537580  ENSMUSG00000003135.15  \n",
       "663344  39538271  ENSMUSG00000003135.15  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "08298f20-f71f-42ad-b949-21331d315e90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e23de0c7-c32f-4461-bb8a-f76ebea93ad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info = pd.read_csv('/Users/fairliereese/mortazavi_lab/data/mousewg/refs/gencode_vM21_gene_metadata.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4b5e8d7c-0277-4a68-b2ca-5e016b9aff10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>length</th>\n",
       "      <th>biotype</th>\n",
       "      <th>biotype_category</th>\n",
       "      <th>tf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>109</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>465597</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000102851.1</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>479</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000103377.1</td>\n",
       "      <td>Gm37180</td>\n",
       "      <td>2818</td>\n",
       "      <td>TEC</td>\n",
       "      <td>other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000104017.1</td>\n",
       "      <td>Gm37363</td>\n",
       "      <td>2232</td>\n",
       "      <td>TEC</td>\n",
       "      <td>other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gid    gname  length               biotype  \\\n",
       "0  ENSMUSG00000064842.1  Gm26206     109                 snRNA   \n",
       "1  ENSMUSG00000051951.5     Xkr4  465597        protein_coding   \n",
       "2  ENSMUSG00000102851.1  Gm18956     479  processed_pseudogene   \n",
       "3  ENSMUSG00000103377.1  Gm37180    2818                   TEC   \n",
       "4  ENSMUSG00000104017.1  Gm37363    2232                   TEC   \n",
       "\n",
       "  biotype_category     tf  \n",
       "0            other  False  \n",
       "1   protein_coding  False  \n",
       "2       pseudogene  False  \n",
       "3            other  False  \n",
       "4            other  False  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "856f253b-f53d-4a63-917d-de981fd60dc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info = gene_info[['gid', 'gname']]\n",
    "gene_info.rename({'gid': 'gene_id', 'gname': 'gene_name'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ea1b58ac-efa1-4297-8973-e47097f09e4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.merge(gene_info, how='left', on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "74e47eb5-855f-4479-b6f8-5187089012c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [Start, Start_tid, End, End_tid]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [Start, Start_tid, End, End_tid]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [Start, Start_tid, End, End_tid]\n",
      "Index: []\n",
      "383100\n",
      "383100\n",
      "Empty DataFrame\n",
      "Columns: [Start, End]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [Start, End]\n",
      "Index: []\n",
      "0\n",
      "0\n",
      "Empty DataFrame\n",
      "Columns: [Start, Start_tid, End, End_tid]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [Start, Start_tid, End, End_tid]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [Start, Start_tid, End, End_tid]\n",
      "Index: []\n",
      "349389\n",
      "349389\n",
      "Empty DataFrame\n",
      "Columns: [Start, End]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [Start, End]\n",
      "Index: []\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "fwd, rev = cerberus.get_stranded_gtf_dfs(df)\n",
    "df = pd.DataFrame()\n",
    "for strand, temp in zip(['+', '-'], [fwd, rev]):\n",
    "    if strand == '+':\n",
    "        start_ind = 0\n",
    "        end_ind = 1\n",
    "    elif strand == '-':\n",
    "        start_ind = 1\n",
    "        end_ind = 0\n",
    "        \n",
    "    exons = pd.DataFrame({'Start':temp['Coord'].iloc[start_ind::2].values,\n",
    "                          'Start_tid':temp['transcript_id'].iloc[start_ind::2].values,\n",
    "                          'End':temp['Coord'].iloc[end_ind::2].values,\n",
    "                          'End_tid':temp['transcript_id'].iloc[end_ind::2].values})\n",
    "    temp = temp.iloc[::2]\n",
    "    temp.drop('Coord', axis=1, inplace=True)\n",
    "\n",
    "    print(exons.loc[exons.End.isnull()])\n",
    "    print(exons.loc[exons.Start.isnull()])\n",
    "\n",
    "    print(exons.loc[exons.Start_tid != exons.End_tid])\n",
    "\n",
    "    exons.drop(['Start_tid','End_tid'], axis=1, inplace=True)\n",
    "\n",
    "    print(len(temp.index))\n",
    "    print(len(exons.index))\n",
    "    # temp[['Start', 'End']] = exons[['Start', 'End']]\n",
    "    temp['Start'] = exons['Start'].tolist()\n",
    "    temp['End'] = exons['End'].tolist()\n",
    "    # print(len(df.index))\n",
    "\n",
    "    print(exons.loc[exons.End.isnull()])\n",
    "    print(exons.loc[exons.Start.isnull()])\n",
    "\n",
    "    print(len(temp.loc[temp.Start.isnull()].index))\n",
    "    print(len(temp.loc[temp.End.isnull()].index))\n",
    "    \n",
    "    df = pd.concat([df, temp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "27f9e3fd-6a50-46fc-a799-ed9b8ae3a358",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add exon feature tag \n",
    "df['Feature'] = 'exon'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "12208a3a-f1ac-4300-b54b-edd109e8513c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name',\n",
    "                   'tss_id', 'tes_id',\n",
    "                   'new_transcript_id', 'original_transcript_id',\n",
    "                   'original_transcript_name', 'ag1', 'ag2']\n",
    "        gb_cols = list(set(gb_cols)&(set(t_df.columns)))\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "\n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "\n",
    "    return t_df\n",
    "\n",
    "def make_gtf_from_exons(df):\n",
    "    \n",
    "    # make transcript entries\n",
    "    t_df = make_hier_entry(df, how='t')\n",
    "    \n",
    "    # make gene entries\n",
    "    g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "    # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "    df = pd.concat([df, t_df, g_df])\n",
    "    df = cerberus.sort_gtf(df)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dcc074f6-bb7d-4a17-af55-6b5396f92982",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, original_transcript_id, Chromosome, Strand, gene_id, gene_name, Start, End, Feature]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Start.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6f15e74f-30f5-4cd5-bc7a-e5f6054eef85",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = make_gtf_from_exons(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ae55400c-17ca-4783-9921-114d04bfe957",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121237217</td>\n",
       "      <td>121255865</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121237217</td>\n",
       "      <td>121254583</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121237217</td>\n",
       "      <td>121237351</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121238419</td>\n",
       "      <td>121238517</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121242501</td>\n",
       "      <td>121242644</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id original_transcript_id Chromosome Strand  \\\n",
       "0                     NaN                    NaN      chr11      +   \n",
       "1  ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "2  ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "3  ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "4  ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "\n",
       "                gene_id gene_name      Start        End     Feature  \n",
       "0  ENSMUSG00000000056.7      Narf  121237217  121255865        gene  \n",
       "1  ENSMUSG00000000056.7      Narf  121237217  121254583  transcript  \n",
       "2  ENSMUSG00000000056.7      Narf  121237217  121237351        exon  \n",
       "3  ENSMUSG00000000056.7      Narf  121238419  121238517        exon  \n",
       "4  ENSMUSG00000000056.7      Narf  121242501  121242644        exon  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b28461ef-f27f-480f-9b35-8160bb262805",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, original_transcript_id, Chromosome, Strand, gene_id, gene_name, Start, End, Feature]\n",
       "Index: []"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.End.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "822cfb74-0f87-4303-845d-83cf6e35cae9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121237217</td>\n",
       "      <td>121254583</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121237217</td>\n",
       "      <td>121237351</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121238419</td>\n",
       "      <td>121238517</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121242501</td>\n",
       "      <td>121242644</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121244594</td>\n",
       "      <td>121244726</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121244936</td>\n",
       "      <td>121245070</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121245338</td>\n",
       "      <td>121245456</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121246926</td>\n",
       "      <td>121247055</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121249129</td>\n",
       "      <td>121249192</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121250293</td>\n",
       "      <td>121250430</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121250760</td>\n",
       "      <td>121250917</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000000056.7_0</td>\n",
       "      <td>ENSMUST00000103015.3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000000056.7</td>\n",
       "      <td>Narf</td>\n",
       "      <td>121252681</td>\n",
       "      <td>121254583</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             transcript_id original_transcript_id Chromosome Strand  \\\n",
       "1   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "2   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "3   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "4   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "5   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "6   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "7   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "8   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "9   ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "10  ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "11  ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "12  ENSMUSG00000000056.7_0   ENSMUST00000103015.3      chr11      +   \n",
       "\n",
       "                 gene_id gene_name      Start        End     Feature  \n",
       "1   ENSMUSG00000000056.7      Narf  121237217  121254583  transcript  \n",
       "2   ENSMUSG00000000056.7      Narf  121237217  121237351        exon  \n",
       "3   ENSMUSG00000000056.7      Narf  121238419  121238517        exon  \n",
       "4   ENSMUSG00000000056.7      Narf  121242501  121242644        exon  \n",
       "5   ENSMUSG00000000056.7      Narf  121244594  121244726        exon  \n",
       "6   ENSMUSG00000000056.7      Narf  121244936  121245070        exon  \n",
       "7   ENSMUSG00000000056.7      Narf  121245338  121245456        exon  \n",
       "8   ENSMUSG00000000056.7      Narf  121246926  121247055        exon  \n",
       "9   ENSMUSG00000000056.7      Narf  121249129  121249192        exon  \n",
       "10  ENSMUSG00000000056.7      Narf  121250293  121250430        exon  \n",
       "11  ENSMUSG00000000056.7      Narf  121250760  121250917        exon  \n",
       "12  ENSMUSG00000000056.7      Narf  121252681  121254583        exon  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id=='ENSMUSG00000000056.7_0']\n",
    "# df.loc[df.transcript_id=='ENSMUSG00000000001.4_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8ee52eed-e730-440a-a4d8-3a6f058ce28b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# decrement the start coordinate to convert bed style coordinates\n",
    "# to gtf style coordinates\n",
    "df['Start'] = df['Start']-1\n",
    "\n",
    "# save example gtfs \n",
    "tids = ['ENSMUSG00000000056.7_0', 'ENSMUSG00000000001.4_0']\n",
    "for t in tids:\n",
    "    o = '{}.gtf'.format(t)\n",
    "    temp = df.loc[df.transcript_id == t]\n",
    "    temp = pr.PyRanges(temp)\n",
    "    temp.to_gtf(o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2dfbac19-e983-4d74-a117-9b2616e1a122",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.PyRanges(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "99be47d5-611e-47c6-b82b-3d0d2697e7f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_gtf(ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd562b6c-7fc7-4fc7-9ce9-f4ce11dc79aa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
