

```{r eval=FALSE}

# load R packages

library(Seurat)
library(tidyverse)
library(cowplot)
library(patchwork)
library(WGCNA)
library(Matrix)
library(viridis)
library(harmony)
library(RColorBrewer)
library(ggpubr)
library(tictoc)
library(RColorBrewer)
library(Hmisc)
library(corrplot)
library(enrichR)
library(GeneOverlap)
library(grid)
library(gridExtra)
library(igraph)
library(ggrepel)
library(hdWGCNA)
enableWGCNAThreads(nThreads = 8)
theme_set(theme_cowplot())
set.seed(12345)

# spatial plotting functions
source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')
source("/pub/smorabit/hdWGCNA/bin/spatial_functions.R")


setwd('/dfs7/swaruplab/smorabit/analysis/scWGCNA/')

# output directories
data_dir <- "data/"
fig_dir <- 'figures/'


color_df <- data.frame(
  colour = c("#454a74",   "#6f9969",   "#5c66a8",   "#808fe1",   "#efc86e" ,  "#97c684"),
  group = c('INH1 VIP+',   'INH2 SST+', 'INH3 PVALB+', 'INH4 LAMP5+',   'INH5 SST+', 'INH6 PVALB+')
)


```

Plot UMAP

```{r eval=FALSE}

plot_list <- PlotEmbedding(
  seurat_obj,
  group.by = 'cell_type',
  split.by = 'Diagnosis',
  raster_dpi = 400,
  raster_scale=0.5, point_size=1,
  plot_theme = umap_theme() + NoLegend(),
  plot_under=TRUE
)

pdf(paste0(fig_dir, 'umap_ASD.pdf'), width=12, height=4)
wrap_plots(plot_list, ncol=3) + plot_layout(guides='collect')
dev.off()


p <- PlotEmbedding(
  seurat_obj,
  group.by = 'annotation',
  selected = c('INH1 VIP+',   'INH2 SST+', 'INH3 PVALB+', 'INH4 LAMP5+',   'INH5 SST+', 'INH6 PVALB+'),
  raster_dpi = 600,
  raster_scale=0.5, point_size=1,
  plot_theme = umap_theme(),
  plot_under=TRUE,
  color_df = color_df
)

pdf(paste0(fig_dir, 'umap_ASD_full.pdf'), width=7, height=7)
p
dev.off()


```

Run hdWGCNA on the INH population

```{r eval=FALSE}


seurat_obj <- SetupForWGCNA(
  seurat_obj,
  gene_select = 'fraction',
  fraction = 0.05,
  group.by = 'annotation',
  wgcna_name = "INH"
)

seurat_obj <- MetacellsByGroups(
  seurat_obj = seurat_obj,
  group.by = c("cell_type", "Sample"),
  ident.group = 'cell_type',
  reduction = 'harmony',
  k = 50,
  max_shared=10,
  target_metacells=250,
  min_cells=50,
  verbose=TRUE
)


seurat_obj <- NormalizeMetacells(seurat_obj)

seurat_obj <- SetDatExpr(
  seurat_obj,
  group_name = "INH",
  group.by='cell_type'
)

seurat_obj <- TestSoftPowers(seurat_obj)
plot_list <- PlotSoftPowers(seurat_obj)

#assemble with patchwork
pdf(paste0(fig_dir, 'softpower.pdf'), width=12, height=8)
wrap_plots(plot_list, ncol=2)
dev.off()


seurat_obj <- ConstructNetwork(seurat_obj, tom_name='INH', overwrite_tom=TRUE)

# plot the dendrogram
pdf(paste0(fig_dir, "inh_dendro.pdf"),height=2, width=4)
PlotDendrogram(seurat_obj, main='INH hdWGCNA Dendrogram')
dev.off()


# compute all MEs in the full single-cell dataset
seurat_obj$Seqbatch <- as.numeric(seurat_obj$Seqbatch)
seurat_obj <- ModuleEigengenes(
  seurat_obj,
  group.by.vars="Seqbatch", # snRNAseq batch
  verbose=FALSE
)

# compute module connectivity:
seurat_obj <- ModuleConnectivity(
  seurat_obj,
  group.by='cell_type',
  group_name = 'INH'
)

################################################################################
# Reset module names and colors
################################################################################

seurat_obj <- ResetModuleNames(
  seurat_obj,
  new_name = "INH-M"
)

library(MetBrewer)
modules <- GetModules(seurat_obj)
mods <- levels(modules$module)
mod_colors <- dplyr::select(modules, c(module, color)) %>%
  distinct %>% arrange(module) %>% .$color
n_colors <- length(mod_colors) -1

new_colors <- paste0(met.brewer("Signac", n=n_colors, type='discrete'))
new_colors <- sample(new_colors)
seurat_obj <- ResetModuleColors(seurat_obj, new_colors)

# save data:
saveRDS(seurat_obj, file=paste0(data_dir, 'Velmeshev_2019_scWGCNA.rds'))


```


Plotting:

```{r eval=FALSE}

########################################
# Feature Plots
########################################

plot_list <- ModuleFeaturePlot(seurat_obj, order=TRUE, raster=TRUE, raster_dpi=400, alpha=1, restrict_range=FALSE, raster_scale=0.25)

plot_list <- lapply(1:length(plot_list), function(x){
  plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))
})

pdf("figures/INH_featureplot_MEs.pdf",height=10, width=5)
wrap_plots(plot_list, ncol=3)
dev.off()


################################################################################
# Hubgene circle plots:
################################################################################

library(igraph)

# individual module networks
ModuleNetworkPlot(
  seurat_obj,
  mods = "all",
  outdir = paste0(fig_dir, 'INH_hubNetworks/')
)


################################################################################
# UMAP:
################################################################################

library(ggrepel)

seurat_obj <- RunModuleUMAP(
  seurat_obj,
  n_hubs = 5,
  n_neighbors=15,
  min_dist=0.3,
  spread=2
)


# label the SFARI genes
sfari_df <- read.csv('~/swaruplab/smorabit/analysis/scWGCNA/ASD_case_study/data/SFARI-Gene_genes_04-29-2022release_07-10-2022export.csv')
sfari_df <- subset(sfari_df, gene.symbol %in% rownames(seurat_obj))


hub_genes <- GetHubGenes(seurat_obj, 25)
label_genes <- intersect(hub_genes$gene_name, sfari_df$gene.symbol)

sfari_hubs <- subset(GetHubGenes(seurat_obj, 50), gene_name %in% sfari_df$gene.symbol)
write.csv(sfari_hubs, quote=FALSE, file=paste0(data_dir, 'SFARI_hubgenes.csv'))

library(reshape2)
library(igraph)
pdf(paste0(fig_dir, 'INH_hubgene_umap_igraph.pdf'), width=10, height=10)
ModuleUMAPPlot(
  seurat_obj,
  edge.alpha=0.5,
  sample_edges=TRUE,
  keep_grey_edges=FALSE,
  edge_prop=0.075, # taking the top 20% strongest edges in each module
  label_genes = label_genes,
  label_hubs=0 # how many hub genes to plot per module?
)
dev.off()


```


Enrichment analysis

```{r eval=FALSE}

library(enrichR)

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021','LINCS_L1000_Chem_Pert_up','LINCS_L1000_Chem_Pert_down', 'WikiPathway_2021_Human', 'KEGG_2021_Human')

# compute GO terms:
seurat_obj <- RunEnrichr(seurat_obj, dbs=dbs)

enrichr_df <- GetEnrichrTable(seurat_obj) %>% subset(P.value < 0.05)
write.table(enrichr_df, quote=FALSE, sep='\t', row.names=FALSE, file=paste0(data_dir, 'ASD_INH_enrichr.tsv'))


# make GO term plots:
EnrichrBarPlot(
  seurat_obj,
  outdir = "figures/INH_enrichr_plots",
  n_terms = 25, plot_size = c(4,16),
  logscale=TRUE
)

# enrichr dotplot
p <- EnrichrDotPlot(
  seurat_obj,
  database = dbs[1], n_terms=3,
  break_ties=TRUE
)
pdf(paste0(fig_dir, 'INH_GO_dotplot.pdf'), width=10, height=10, useDingbats=FALSE)
p
dev.off()

# inspect the enrichr table:
enrichr_df <- GetEnrichrTable(seurat_obj)


```


Control vs Condition stacked violin plots

```{r eval=FALSE}

# seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)
MEs <- GetMEs(seurat_obj)
mods <- colnames(MEs); mods <- mods[mods != 'grey']
seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)

library(ggpubr)

p <- custom_vln(
    seurat_obj,
    features = mods,
    group.by = 'annotation',
    groups = c('INH1 VIP+', 'INH4 LAMP5+', 'INH2 SST+',  'INH5 SST+',  'INH3 PVALB+', 'INH6 PVALB+'),
    add_boxplot=FALSE,
    split.by = 'Diagnosis',
    selected_split = c('ASD', 'Control'),
    split_colors=c('darkorchid', 'grey'),
    add_colorbar=FALSE,
    plot_ymin = NA
  )


pdf(paste0(fig_dir, 'INH_hME_vln_stack.pdf'), width=5, height=8)
p
dev.off()


```


Overlap with SFARI genes:

```{r eval=FALSE}

library(GeneOverlap)

# load SFARI genes
sfari_df <- read.csv('~/swaruplab/smorabit/analysis/scWGCNA/ASD_case_study/data/SFARI-Gene_genes_04-29-2022release_07-10-2022export.csv')

sfari_df <- subset(sfari_df, gene.symbol %in% rownames(seurat_obj))



# load modules
modules <- GetModules(seurat_obj)
mods <- levels(modules$module)
genome.size <- nrow(modules)


overlap_df <- do.call(rbind, lapply(mods, function(cur_mod){

  cur_genes <- modules %>% subset(module == cur_mod) %>% .$gene_name

  cur_overlap <- testGeneOverlap(newGeneOverlap(
      cur_genes,
      sfari_df$gene.symbol,
      genome.size=genome.size
  ))

  cur_overlap <- data.frame(
    'odds.ratio' = cur_overlap@odds.ratio,
    'pval' = cur_overlap@pval,
    'Jaccard' = cur_overlap@Jaccard,
    'size_intersection' = length(cur_overlap@intersection),
    'module' = cur_mod
  )

  cur_overlap

})) %>% as.data.frame()

overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))
overlap_df <- overlap_df %>% subset(module != 'grey')

################################################################################
# Plot as a lollipop
################################################################################

overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)
overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)
overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))

mod_colors <- dplyr::select(modules, c(module, color)) %>%
  distinct
cp <- mod_colors$color; names(cp) <- mod_colors$module

p <- overlap_df %>%
  ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +
  geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +
  geom_point() +
  geom_point(shape=overlap_df$shape, color='black', fill=NA) +
  scale_color_manual(values=cp, guide='none') +
  ylab('') + xlab("Odds ratio") +
  scale_x_continuous(breaks = c(0, 1, 2,3)) +
  labs(size='Size\nintersection') +
  ggtitle('Overlap with SFARI genes') +

  theme(
    panel.border = element_rect(size=1, color='black', fill=NA),
    axis.line.y = element_blank(),
    axis.line.x = element_blank(),
    plot.title = element_text(hjust=0.5, face='plain')
  )

pdf(paste0(fig_dir, 'sfari_overlap2.pdf'), width=4, height=3.5)
p
dev.off()


```

Differential module eigengene (DME) analysis comparing ASD & Control in each INH cluster

```{r eval=FALSE}

group_col <- 'annotation'
groups <- c('INH1 VIP+', 'INH4 LAMP5+', 'INH2 SST+',  'INH5 SST+',  'INH3 PVALB+', 'INH6 PVALB+')
DME_list <- list()
for(cur_group in groups){
  print(cur_group)
  # get cell barcodes for this pseudotime bin in control & AD
  g1 <- seurat_obj@meta.data[seurat_obj[[group_col]] == cur_group & seurat_obj$Diagnosis == 'ASD',] %>% rownames
  g2 <- seurat_obj@meta.data[seurat_obj[[group_col]] == cur_group & seurat_obj$Diagnosis == 'Control',] %>% rownames

  DMEs <- FindDMEs(
    seurat_obj,
    barcodes1 = g1,
    barcodes2 = g2,
    test.use='wilcox',
    wgcna_name='INH2',
    harmonized=TRUE
  )
  DMEs$group <- cur_group

  # fix infs:
  DMEs$avg_log2FC <- ifelse(abs(DMEs$avg_log2FC) == Inf, 0, DMEs$avg_log2FC)
  DME_list[[cur_group]] <- DMEs

}
DMEs <- do.call(rbind, DME_list)
DMEs$group <- factor(
  as.character(DMEs$group),
  levels = groups
)


write.csv(DMEs, row.names=FALSE, quote=FALSE, file=paste0(data_dir, "ASD_INH_DMEs.csv"))

```



Module Preservation with Nagy et al dataset

```{r eval=FALSE}


seurat_query <- readRDS('~/swaruplab/smorabit/analysis/scWGCNA/data/Nagy_2020.rds')

p <- PlotEmbedding(
  seurat_query,
  group.by = 'annotation',
  raster_dpi = 600,
  raster_scale=0.5, point_size=1,
  plot_theme = umap_theme(),
  plot_under=TRUE
)

pdf(paste0(fig_dir, 'umap_nagy.pdf'), width=7, height=7)
p
dev.off()

seurat_query <- ProjectModules(
  seurat_obj = seurat_query,
  seurat_ref = seurat_obj,
  wgcna_name = "INH",
  wgcna_name_proj="INH_projected"
)

plot_list <- ModuleFeaturePlot(seurat_query, order=TRUE, raster=TRUE, raster_dpi=400, alpha=1, restrict_range=FALSE, raster_scale=0.25)
wrap_plots(plot_list, ncol=4)
dev.off()

plot_list <- lapply(1:length(plot_list), function(x){
  plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))
})

pdf("figures/INH_featureplot_MEs_Nagy.pdf",height=6, width=7)
wrap_plots(plot_list, ncol=5)
dev.off()

# run module preservation function for control & AD MG separately
seurat_query <- SetDatExpr(
  seurat_query,
  group_name = "INH",
  group.by = "cell_type",
  use_metacells = FALSE,
  assay = 'RNA'
)

seurat_query <- ModulePreservation(
  seurat_query,
  seurat_ref = seurat_obj,
  name="INH_projected",
  wgcna_name = 'INH_projected',
  wgcna_name_ref = 'INH2',
  verbose=3,
  n_permutations=100
)

library(ggrepel)
plot_list <- PlotModulePreservation(
  seurat_query,
  name="INH_projected",
  statistics = "summary"
)

pdf(paste0(fig_dir, 'INH_module_preservation_summary_Nagy.pdf'), width=8, height=4)
wrap_plots(plot_list, ncol=2)
dev.off()

saveRDS(seurat_query, file=paste0(data_dir, 'Nagy_2020_hdWGCNA.rds'))



```

Plot selected GO terms:

```{r eval=FALSE}

modules <- GetModules(seurat_obj)
color_df <- modules %>% subset(module!='grey') %>%
  select(c(module, color)) %>% distinct %>%
  mutate(module=droplevels(module)) %>%
  rename(c(group=module, colour=color))
mods <- levels(modules$module); mods <- mods[mods != 'grey']

color_df$group <- factor(as.character(color_df$group), levels=mods)

# helper function to wrap text
wrapText <- function(x, len) {
    sapply(x, function(y) paste(strwrap(y, len), collapse = "\n"), USE.NAMES = FALSE)
}

combined_output <- GetEnrichrTable(seurat_obj)
selected_terms <- read.delim('data/ASD_INH_enrichr_selected.txt', sep='\t', header=1)

# subset selected terms
selected_terms <- subset(combined_output, Term %in% selected_terms$Term & P.value < 0.05)

selected_terms$group <- factor(
  as.character(selected_terms$module),
  levels = mods
)

# set max pval

quantile(-log(selected_terms$P.value), 0.95)
max_p <- 10

selected_terms$logp <- -log(selected_terms$P.value)
selected_terms$logp <- ifelse(selected_terms$logp > max_p, max_p, selected_terms$logp)

# remove GO Term ID
selected_terms$Term <- str_replace(selected_terms$Term, " \\s*\\([^\\)]+\\)", "")

selected_terms <- selected_terms %>%
  arrange(group)


selected_terms$wrap <- wrapText(selected_terms$Term, 35)

selected_terms$Term <- factor(
  as.character(selected_terms$Term),
  levels = rev(unique(as.character(selected_terms$Term)))
)

# GO Term dot plot

p <- selected_terms %>%
  ggplot(aes(x = group, y = Term, color =logp, size=log(Combined.Score))) +
  geom_point() +
  scale_color_stepsn(colors=rev(magma(256))) +
  RotatedAxis() + xlab('') + ylab('') +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    panel.border = element_rect(size=1, color='black', fill=NA),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    plot.margin = margin(0,0,0,0),
    panel.grid = element_line(size=0.25, color='lightgrey')
  )


# make the colorbar as its own heatmap
color_df$var <- 1
cp <- color_df$colour; names(cp) <- color_df$group
colorbar <- color_df %>%
  ggplot(aes(x=group, y=var, fill=group)) +
  geom_tile() +
  scale_fill_manual(values=cp) +
  coord_equal() +
  NoLegend() + RotatedAxis() +
  theme(
    plot.title=element_blank(),
    axis.line=element_blank(),
    axis.ticks.y =element_blank(),
    axis.text.y = element_blank(),
    axis.title = element_blank(),
    plot.margin=margin(0,0,0,0),
  )




pdf(paste0(fig_dir, 'ASD_selected_GO_terms.pdf'), width=9, height=6)
p / colorbar #+ plot_layout(heights=c(20,1))
dev.off()

library(viridis)


```
